docs/architecture/grpc-for-wcf-developers/apm.md

```diff
@@ -1,11 +1,278 @@
-description: TO BE WRITTEN
+description: Logging, metrics and tracing for ASP.NET Core gRPC applications
 author: markrendle
 ms.date: 09/02/2019
 ---
 
 # Application Performance Management
 
+In modern production environments like Kubernetes, it is very important to be able to monitor applications to ensure they are running optimally. Concerns like logging and metrics have never been more important. ASP.NET Core, including gRPC, has first-class support for producing and managing log messages and metrics data, as well as *tracing* data. This section will explore these areas in more details.
+
+## Logging vs Metrics
+
+Before looking at the implementation details, it is necessary to understand the difference between logging and metrics.
+
+Logging is concerned with text messages that record detailed information about things that have happened in the system. Log messages may include exception data like stack traces, or structured data that provides context about the message. Logging output is commonly written to a searchable text store.
+
+Metrics refers to numeric data that is designed to be aggregated and presented using charts and graphs in a dashboard that provides a view of the overall health and performance of an application. Metrics data can also be used to trigger automated alerts when a threshold is exceeded. Examples of metrics data include the time taken to process requests, the number of requests per second being handled by an instance of a service, or the number of failed requests on an instance.
+
+## Logging in ASP.NET Core gRPC
+
+ASP.NET Core provides support for logging out-of-the-box, in the form of the [Microsoft.Extensions.Logging package](https://www.nuget.org/packages/Microsoft.Extensions.Logging). This is included with the Web SDK so there is no need to install it manually.
+
+### Producing log messages
+
+The Logging extension is automatically registered with ASP.NET Core's dependency injection system, so loggers can be specified as a constructor parameter on gRPC service types.
+
+```csharp
+public class StockData : Stocks.StocksBase
+{
+    private readonly ILogger<StockData> _logger;
+
+    public StockData(ILogger<StockData> logger)
+    {
+        _logger = logger;
+    }
+}
+```
+
+The `ILogger<T>` interface is generic so that it can use the fully-qualified name of the class that generated the message as the *Log Category*. There is a non-generic `ILogger` interface that can be created with a custom *Log Category* string, but using the generic version means that categories are consistently applied throughout an application.
+
+Log messages are assigned *levels* indicating their importance, which allows the components writing the log output to various destinations whether to include messages or not. These levels allow developers to generate more log messages while debugging or testing code, and fewer messages in production systems to avoid overwhelming the storage providers.
```

Comment:
IMO this section on log levels could be shortened and replaced with links to existing logging documentation.